### PR TITLE
libs/os: add test case for TrapSignal

### PR DESCRIPTION
**Pull request by alessio**
Add test case for TrapSignal.

### CODE REVIEW

**Comment by alessio** [comment]
I'm trying to move `TrapSignal` out of the `cosmos-sdk` as IMvHO `github.com/tendermint/tendermint/libs/os` is the right place where it should live.

**Comment by melekes** [comment]
I am no expert, but the reason we return zero is https://github.com/tendermint/tendermint/issues/3238. We should probably discuss with @mdyring before making this change.

**Comment by mdyring** [comment]
> I am no expert, but the reason we return zero is #3238. We should probably discuss with @mdyring before making this change.

I am not aware how compliant systemd is with [1]. Off the top of my head it seems like a sensible change @alessio is suggesting, but maybe give it a spin first and make sure it doesn't break for everyone using systemd. :-)

**Comment by alessio** [comment]
Please never use "systemd" and "compliant" in the same sentence - it's an oxymoron @mdyring :)

From what I can [read here](https://freedesktop.org/software/systemd/man/systemd.exec.html#id-1.20.8), I'm opmimistic though. At least wrt exit codes handling, Poettering and his team seems to have avoided reinventing the wheel.

**Comment by erikgrinaker** [comment]
I believe exit code >128 is when the program doesn't handle the signal and dies because of it. If the program traps e.g. `SIGTERM` and does a clean shutdown, this is a successful exit and should have exit code `0`.

https://stackoverflow.com/questions/39269536/what-should-my-programs-exit-code-be-if-i-caught-a-signal

**Comment by erikgrinaker** [comment]
When in doubt, do what Postgres does :)

```
$ /usr/local/opt/postgresql/bin/postgres -D /usr/local/var/postgres
2020-11-11 11:41:44.706 CET [13229] LOG:  starting PostgreSQL 12.4 on x86_64-apple-darwin19.5.0, compiled by Apple clang version 11.0.3 (clang-1103.0.32.62), 64-bit

$ kill -TERM 133229
2020-11-11 11:41:56.191 CET [13229] LOG:  received smart shutdown request
2020-11-11 11:41:56.192 CET [13229] LOG:  background worker "logical replication launcher" (PID 13236) exited with exit code 1
2020-11-11 11:41:56.192 CET [13231] LOG:  shutting down
2020-11-11 11:41:56.197 CET [13229] LOG:  database system is shut down
$ echo $?
0
```

**Comment by alessio** [comment]
> When in doubt, do what Postgres does :)

Postgres catches the signal, handles it and exits with 0. If a cleanup function is passed as an argument and executed, that makes sense to me. This is a call on the semantic here, I'm happy to amend the PR accordingly!

@erikgrinaker any comment on the testcase? Happy to change it if necessary

**Comment by erikgrinaker** [comment]
Since the purpose of `TrapSignal` is to trap specifically `SIGTERM` and `SIGINT` and run a cleanup function, I think it should always return 0.

However, if the program dies as a consequence of some other signal (e.g. `SIGKILL`) then it should return >128. I believe the OS already does this for us by default, e.g.:

```
$ sleep 60

($ kill -9 13716)
Killed: 9
$ echo $?
137
```

**Comment by alessio** [comment]
Yep, SIGKILL along with other signals (e.g SIGABRT) cannot be trapped.

Fair enough - I'll revert the changes and leave only the test around if you reckon it adds any value.

**Comment by erikgrinaker** [comment]
Sure, tests are always good! 👍

**Review by erikgrinaker — APPROVED** [review]

File: libs/os/os.go [hunk 1]
```diff
@@ -13,19 +13,19 @@ type logger interface {
...
 	signal.Notify(c, os.Interrupt, syscall.SIGTERM)
+
 	go func() {
-		for sig := range c {
-			logger.Info(fmt.Sprintf("captured %v, exiting...", sig))
-			if cb != nil {
-				cb()
-			}
-			os.Exit(0)
+		sig := <-c
```
Comment (alessio) [inline libs/os/os.go:23]: I've removed the for loop as this is enough to block. Plus, decreasing indentation slightly improves readability.

**Review by melekes — APPROVED** [review]
👍

**Pull request merged**